### PR TITLE
Optimize collateral auction

### PR DESCRIPTION
**Pull request by wangjj9219**
if collateral's target is not zero, when auction end, exists following situations:
1. if dex can provider same for target, dex take the auction
2. if the last bid price reach target and higher than dex, the auction dealt by last bidder
3. other, auction aborted.

if collateral's target is zero, when auction end, exists following situations:
1. if dex can provider same or more last bid price, dex take the auction.
2. if exists bid, the auction dealt by last bidder.
3. if no bid, auction aborted.

### CODE REVIEW

File: modules/auction-manager/src/lib.rs [hunk 1]
```diff
@@ -584,98 +594,118 @@ impl<T: Config> Pallet<T> {
...
+					refund_collateral,
+				);
+				if let Err(e) = res {
+					log::warn!(
+						target: "auction-manager",
+						"withdraw_collateral: failed to withdraw {:?} {:?} from CDP treasury to {:?}: {:?}. \
+						This is unexpected but should be safe",
+						refund_collateral, collateral_auction.currency_id, collateral_auction.refund_recipient, e
+					);
+					debug_assert!(false);
```
Comment (xlc) [inline modules/auction-manager/src/lib.rs:632]: this should be a helper method

File: modules/auction-manager/src/lib.rs [hunk 2]
```diff
@@ -584,98 +594,118 @@ impl<T: Config> Pallet<T> {
...
 			Self::deposit_event(Event::CollateralAuctionDealt {
 				auction_id,
 				collateral_type: collateral_auction.currency_id,
 				collateral_amount: collateral_auction.amount,
-				winner: bidder,
+				winner,
 				payment_amount,
 			});
+		} else {
+			// refund stable to the last_bidder, it shouldn't fail and affect the process.
```
Comment (xlc) [inline modules/auction-manager/src/lib.rs:686]: need some explanation on what happened here

**Review by xlc — COMMENTED** [review]
not 100% sure if we want to abort the auction or just restart it
maybe we should only abort it if we already have too many ongoing auctions?

**Review by shaunxw — COMMENTED** [review]
Some nitpicks.

I think we can introduce some restart mechanism if auction aborted, but it could be future improvements.

**Review by shaunxw — APPROVED** [review]
LGTM, will need to fix CI errors.

**Review by xlc — APPROVED** [review]

**Review by shaunxw — APPROVED** [review]

**Pull request merged**